llvm/include/llvm/MC/DXContainerRootSignature.h

```diff
@@ -6,23 +6,34 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "llvm/BinaryFormat/DXContainer.h"
 #include <cstdint>
 #include <limits>
 
 namespace llvm {
 
 class raw_ostream;
-
 namespace mcdxbc {
+
+struct RootParameter {
+  dxbc::RootParameterHeader Header;
+  union {
+    dxbc::RootConstants Constants;
+  };
+};
 struct RootSignatureDesc {
-  uint32_t Version = 2;
-  uint32_t NumParameters = 0;
-  uint32_t RootParametersOffset = 0;
-  uint32_t NumStaticSamplers = 0;
-  uint32_t StaticSamplersOffset = 0;
-  uint32_t Flags = 0;
+
+  dxbc::RootSignatureHeader Header;
+  SmallVector<mcdxbc::RootParameter> Parameters;
+  RootSignatureDesc()
+      : Header(dxbc::RootSignatureHeader{
+            2, 0, sizeof(dxbc::RootSignatureHeader), 0, 0, 0}) {}
 
   void write(raw_ostream &OS) const;
+
+  size_t getSize() const {
+    return sizeof(dxbc::RootSignatureHeader) + Parameters.size_in_bytes();
+  }
```

Comment:
This doesn't look right. You're putting `Constants` in a union above to define a root parameter, but the root parameters other than Constants won't have the same size as the constants do. This would imply that the size of the `RootParameter` struct will be the maximum size of any root parameter, so `Parameters.size_in_bytes()` won't match the serialized size at all here.